src/data/java/com/fabriccommunity/spookytime/SpookyRecipeProvider.java

```diff
@@ -0,0 +1,139 @@
+	{
+		ShapedRecipeJsonFactory.create(Items.DIAMOND)
+				.pattern("   ")
+				.pattern("   ")
+				.pattern("  b")
+				.input('b', Items.DIAMOND)
+				.criterion("has_diamond", hasItem(Items.DIAMOND))
+				.offerTo(consumer, SpookyTime.id("example_recipe"));
+	}
+
+	@Override
+	public String getName()
+	{
+		return "Spooky Recipe Provider";
+	}
+
+	@Override
+	public void run(DataCache dataCache_1)
+	{
+		Path path_1 = this.root.getOutput();
+		Set<Identifier> set_1 = Sets.newHashSet();
+		this.generate((recipeJsonProvider_1) ->
+		{
+			if (!set_1.add(recipeJsonProvider_1.getRecipeId()))
+			{
+				throw new IllegalStateException("Duplicate recipe " + recipeJsonProvider_1.getRecipeId());
+			} else
+			{
+				this.writeRecipeJson(dataCache_1, recipeJsonProvider_1.toJson(), path_1.resolve("data/" + recipeJsonProvider_1.getRecipeId().getNamespace() + "/recipes/" + recipeJsonProvider_1.getRecipeId().getPath() + ".json"));
+				JsonObject jsonObject_1 = recipeJsonProvider_1.toAdvancementJson();
+				if (jsonObject_1 != null)
+				{
+					this.generateProgressionAdvancements(dataCache_1, jsonObject_1, path_1.resolve("data/" + recipeJsonProvider_1.getRecipeId().getNamespace() + "/advancements/" + recipeJsonProvider_1.getAdvancementId().getPath() + ".json"));
+				}
+
+			}
+		});
+	}
+
+	private void writeRecipeJson(DataCache dataCache_1, JsonObject jsonObject_1, Path path_1)
+	{
+		try
+		{
+			String string_1 = GSON.toJson(jsonObject_1);
+			String string_2 = SHA1.hashUnencodedChars(string_1).toString();
+			if (!Objects.equals(dataCache_1.getOldSha1(path_1), string_2) || !Files.exists(path_1))
+			{
+				Files.createDirectories(path_1.getParent());
+				try (BufferedWriter bufferedWriter_1 = Files.newBufferedWriter(path_1))
+				{
+					bufferedWriter_1.write(string_1);
+				}
+			}
+
+			dataCache_1.updateSha1(path_1, string_2);
+		} catch (IOException var19)
+		{
+			LOGGER.error("Couldn't save recipe {}", path_1, var19);
+		}
+
+	}
+
+	private void generateProgressionAdvancements(DataCache dataCache_1, JsonObject jsonObject_1, Path path_1)
+	{
+		try
+		{
+			String string_1 = GSON.toJson(jsonObject_1);
+			String string_2 = SHA1.hashUnencodedChars(string_1).toString();
+			if (!Objects.equals(dataCache_1.getOldSha1(path_1), string_2) || !Files.exists(path_1))
+			{
+				Files.createDirectories(path_1.getParent());
+				try (BufferedWriter bufferedWriter_1 = Files.newBufferedWriter(path_1))
+				{
+					bufferedWriter_1.write(string_1);
+				}
+			}
+
+			dataCache_1.updateSha1(path_1, string_2);
+		} catch (IOException var19)
+		{
+			LOGGER.error("Couldn't save recipe advancement {}", path_1, var19);
+		}
+
+	}
+
+	private CriterionConditions hasItem(ItemConvertible item)
+	{
+		return hasItem(item, 1);
+	}
+
+	private CriterionConditions hasItem(ItemConvertible item, int count)
+	{
+		return InventoryChangedCriterion.Conditions.items(ItemPredicate.Builder.create().item(item).count(NumberRange.IntRange.atLeast(count)).build());
+	}
+}
```

Comment:
This whole thing looks... unwieldy and unnecessary. I don't see why we can't just use JSONs.